tools/lkl/lib/rump-host.c

```diff
@@ -0,0 +1,672 @@
+					    td->timeout.tv_nsec);
+		if (td->canceled) {
+			if (!td->thrid)
+				rumpuser_free(td, 0);
+			goto end;
+		}
+		rumpuser_mutex_exit(td->mtx);
+		/* FIXME: we should not use rumpuser__errtrans here
+		 * 60==ETIMEDOUT(netbsd), rumpuser__errtrans(ETIMEDOUT))
+		 */
+		if (err && err != 60)
+			goto end;
+	}
+
+	f(thrarg);
+
+	rumpuser_thread_exit();
+end:
+	return arg;
+}
+
+static void rump_timer_cancel(void *timer)
+{
+	struct thrdesc *td = timer;
+
+	if (td->canceled)
+		return;
+
+	td->canceled = 1;
+	rumpuser_mutex_enter(td->mtx);
+	rumpuser_cv_signal(td->cv);
+	rumpuser_mutex_exit(td->mtx);
+
+	rumpuser_mutex_destroy(td->mtx);
+	rumpuser_cv_destroy(td->cv);
+
+	if (td->thrid)
+		rumpuser_thread_join(td->thrid);
+
+	rumpuser_free(td, 0);
+}
+
+/* from src-netbsd/sys/rump/librump/rumpkern/thread.c */
+static void rump_thread_allow(struct lwp *l)
+{
+	rumpuser_mutex_enter(thrmtx);
+	if (l == NULL)
+		threads_are_go = true;
+
+	rumpuser_cv_broadcast(thrcv);
+	rumpuser_mutex_exit(thrmtx);
+}
+
+static unsigned long long time_ns(void)
+{
+	struct timespec ts;
+
+	rumpuser_clock_gettime(RUMPUSER_CLOCK_RELWALL, (int64_t *)&ts.tv_sec,
+			       &ts.tv_nsec);
+
+	return ((unsigned long long) ts.tv_sec * NSEC_PER_SEC) + ts.tv_nsec;
+}
+
+static void *timer_alloc(void (*fn)(void *), void *arg)
+{
+	struct thrdesc *td;
+
+	rumpuser_malloc(sizeof(*td), 0, (void **)&td);
+
+	memset(td, 0, sizeof(*td));
+	td->f = fn;
+	td->arg = arg;
+
+	rumpuser_mutex_init(&td->mtx, RUMPUSER_MTX_SPIN);
+	rumpuser_cv_init(&td->cv);
+
+	return td;
+}
+
+static int timer_set_oneshot(void *_timer, unsigned long ns)
+{
+	int ret;
+	struct thrdesc *td = _timer;
+
+	td->timeout = (struct timespec){ .tv_sec = ns / NSEC_PER_SEC,
+					 .tv_nsec = ns % NSEC_PER_SEC};
+	ret = rumpuser_thread_create(rump_timer_trampoline, td, "timer",
+				     0, 0, -1, &td->thrid);
+
+	return ret ? -1 : 0;
+}
+
+static void timer_free(void *_timer)
+{
+	rump_timer_cancel(_timer);
+}
+
+static void panic(void)
+{
+	rumpuser_exit(RUMPUSER_PANIC);
+}
+
+struct lkl_host_operations lkl_host_ops = {
+	.panic = panic,
+	.thread_create = rump_thread_create,
+	.thread_detach = rump_thread_detach,
+	.thread_exit = rump_thread_exit,
+	.thread_join = rump_thread_join,
+	.thread_self = rump_thread_self,
+	.thread_equal = rump_thread_equal,
+	.sem_alloc = rump_sem_alloc,
+	.sem_free = rump_sem_free,
+	.sem_up = rump_sem_up,
+	.sem_down = rump_sem_down,
+	.mutex_alloc = rump_mutex_alloc,
+	.mutex_free = rump_mutex_free,
+	.mutex_lock = rump_mutex_lock,
+	.mutex_unlock = rump_mutex_unlock,
+	.tls_alloc = rump_tls_alloc,
+	.tls_free = rump_tls_free,
+	.tls_set = rump_tls_set,
+	.tls_get = rump_tls_get,
+	.time = time_ns,
+	.timer_alloc = timer_alloc,
+	.timer_set_oneshot = timer_set_oneshot,
+	.timer_free = timer_free,
+	.print = rump_print,
+	.mem_alloc = rump_mem_alloc,
+	.mem_free = rump_mem_free,
+	.ioremap = lkl_ioremap,
+	.iomem_access = lkl_iomem_access,
+	.jmp_buf_set = jmp_buf_set,
```

Comment:
I know this is the LKL way of doing this.  Perhaps this isn't the place to ask this, but why not include a more abstract function that simply switches to another thread.  Could be implemented with setjmp/longjmp, or with OS primitives below this layer.

Reply:
(asking these question here should be fine with me)

I'm not sure if I understand your question.  maybe the commit https://github.com/libos-nuse/lkl-linux/commit/79131f214e2730f3e0cc7d290f0e739ac931a845 is relevant to this question ?

Reply:
@gparmer I have an old branch that removes most of this low level ops and switches to higher level opts like alloc/switch/free contexts, see: https://github.com/tavip/linux/commits/the-expanse, in particular https://github.com/tavip/linux/commit/e182fe0e48f1f7f1ed8dc31eed68dddc19a6e6cf.

The reason I did not merge this is that we have optimizations to reduce the system call latency (see e1cd8395b36e951d4b1dc06395b57aae03ea3708) which I could not map well with the new scheme.